### ISSUE TITLE
Flexibility in choosing which checkpoint to save during multi-head training.

### ISSUE THREAD

**Issue by LarsSchaaf** [ACEsuit/mace#479]
**Current behaviour**

If the errors on the mp-head increase and the default-head decrease the overall loss may increases, meaning that no checkpoint is saved. This means that eg. after 200 epochs of training, where the error on the fine-tuning dataset decreases significantly the last saved checkpoint could be eg. epoch 8. 

**Desired behaviour**

Choose to save and select the final model based on the loss of a desired head (or a weighting of multiple heads). Such that if the error decreases on the finetuning validation set, I save this model. 

**Current workaround**

Save checkpoints after each epoch.

**Comment by ilyes319** [comment]
@LarsSchaaf Good point, I made a change yesterday to save only based on the last head (which is the finetuning head) loss but I agree more flexibility should be welcomed.

**Comment by jungsdao** [comment]
Is this problem solved already? I'm having the same issue as well.

**Comment by ilyes319** [comment]
currently only the default validation head is taken as validation for checkpoint. I recommend using the save_all option if you want to pick your own checkpoints.

**Comment by LarsSchaaf** [comment]
**Solution:** for full control of which model to save, use `--save_all_checkpoints` as described in the workaround of the initial issue.

**Comment by jungsdao** [comment]
With `--save_all_checkpoints`, I can save all checkpoint file from various epochs but what eventually used for making a `*.model` is from the last epoch, not from the best one. Is it right? If that's the case, perhaps this problem might need to be addressed?

**Comment by LarsSchaaf** [comment]
What's the desired behaviour would you say? My understanding is that currently the checkpoint with the lowest loss on the fine-tuning head is saved.

**Comment by jungsdao** [comment]
This is the last part of my training log file and the lowest loss func value is from epoch 1182 but what is actually saved as `*_swa.model` seems to be from epoch 1198. 

```
2024-07-10 13:00:55.128 INFO: head: Default, Epoch 1174: loss=0.1852, RMSE_E_per_atom=2.0 meV, RMSE_F=90.8 meV / A, RMSE_stress=5.8 meV / A^3
2024-07-10 13:01:14.248 INFO: head: pbe_mp, Epoch 1176: loss=1.7983, RMSE_E_per_atom=118.9 meV, RMSE_F=193.5 meV / A, RMSE_stress=20.2 meV / A^3
2024-07-10 13:01:14.351 INFO: head: Default, Epoch 1176: loss=0.1877, RMSE_E_per_atom=3.6 meV, RMSE_F=91.1 meV / A, RMSE_stress=5.8 meV / A^3
2024-07-10 13:01:33.313 INFO: head: pbe_mp, Epoch 1178: loss=1.7924, RMSE_E_per_atom=118.7 meV, RMSE_F=193.5 meV / A, RMSE_stress=20.2 meV / A^3
2024-07-10 13:01:33.421 INFO: head: Default, Epoch 1178: loss=0.1870, RMSE_E_per_atom=2.8 meV, RMSE_F=91.1 meV / A, RMSE_stress=5.8 meV / A^3
2024-07-10 13:01:52.724 INFO: head: pbe_mp, Epoch 1180: loss=1.7999, RMSE_E_per_atom=118.9 meV, RMSE_F=194.1 meV / A, RMSE_stress=20.2 meV / A^3
2024-07-10 13:01:52.829 INFO: head: Default, Epoch 1180: loss=0.1834, RMSE_E_per_atom=2.9 meV, RMSE_F=90.5 meV / A, RMSE_stress=5.8 meV / A^3
2024-07-10 13:02:11.896 INFO: head: pbe_mp, Epoch 1182: loss=1.8011, RMSE_E_per_atom=119.0 meV, RMSE_F=193.8 meV / A, RMSE_stress=20.3 meV / A^3
2024-07-10 13:02:12.001 INFO: head: Default, Epoch 1182: loss=0.1833, RMSE_E_per_atom=3.3 meV, RMSE_F=90.5 meV / A, RMSE_stress=5.8 meV / A^3
2024-07-10 13:02:31.300 INFO: head: pbe_mp, Epoch 1184: loss=1.8160, RMSE_E_per_atom=119.5 meV, RMSE_F=194.9 meV / A, RMSE_stress=20.6 meV / A^3
2024-07-10 13:02:31.403 INFO: head: Default, Epoch 1184: loss=0.1858, RMSE_E_per_atom=3.8 meV, RMSE_F=90.8 meV / A, RMSE_stress=5.8 meV / A^3
2024-07-10 13:02:50.106 INFO: head: pbe_mp, Epoch 1186: loss=1.8085, RMSE_E_per_atom=119.2 meV, RMSE_F=194.8 meV / A, RMSE_stress=20.2 meV / A^3
2024-07-10 13:02:50.212 INFO: head: Default, Epoch 1186: loss=0.1841, RMSE_E_per_atom=2.5 meV, RMSE_F=90.9 meV / A, RMSE_stress=5.8 meV / A^3
2024-07-10 13:03:09.036 INFO: head: pbe_mp, Epoch 1188: loss=1.7930, RMSE_E_per_atom=118.8 meV, RMSE_F=193.4 meV / A, RMSE_stress=20.4 meV / A^3
2024-07-10 13:03:09.139 INFO: head: Default, Epoch 1188: loss=0.1835, RMSE_E_per_atom=2.1 meV, RMSE_F=90.9 meV / A, RMSE_stress=5.8 meV / A^3
2024-07-10 13:03:27.967 INFO: head: pbe_mp, Epoch 1190: loss=1.7888, RMSE_E_per_atom=118.6 meV, RMSE_F=193.6 meV / A, RMSE_stress=20.6 meV / A^3
2024-07-10 13:03:28.068 INFO: head: Default, Epoch 1190: loss=0.1878, RMSE_E_per_atom=2.0 meV, RMSE_F=91.3 meV / A, RMSE_stress=5.8 meV / A^3
2024-07-10 13:03:46.921 INFO: head: pbe_mp, Epoch 1192: loss=1.8058, RMSE_E_per_atom=119.3 meV, RMSE_F=193.7 meV / A, RMSE_stress=20.6 meV / A^3
2024-07-10 13:03:47.023 INFO: head: Default, Epoch 1192: loss=0.1930, RMSE_E_per_atom=1.8 meV, RMSE_F=91.7 meV / A, RMSE_stress=5.8 meV / A^3
2024-07-10 13:04:05.914 INFO: head: pbe_mp, Epoch 1194: loss=1.8182, RMSE_E_per_atom=119.7 meV, RMSE_F=193.9 meV / A, RMSE_stress=20.4 meV / A^3
2024-07-10 13:04:06.016 INFO: head: Default, Epoch 1194: loss=0.1942, RMSE_E_per_atom=2.1 meV, RMSE_F=91.9 meV / A, RMSE_stress=5.7 meV / A^3
2024-07-10 13:04:24.870 INFO: head: pbe_mp, Epoch 1196: loss=1.8039, RMSE_E_per_atom=119.2 meV, RMSE_F=193.6 meV / A, RMSE_stress=20.4 meV / A^3
2024-07-10 13:04:24.972 INFO: head: Default, Epoch 1196: loss=0.1906, RMSE_E_per_atom=1.7 meV, RMSE_F=91.0 meV / A, RMSE_stress=5.8 meV / A^3
2024-07-10 13:04:43.811 INFO: head: pbe_mp, Epoch 1198: loss=1.8039, RMSE_E_per_atom=119.0 meV, RMSE_F=194.7 meV / A, RMSE_stress=20.3 meV / A^3
2024-07-10 13:04:43.914 INFO: head: Default, Epoch 1198: loss=0.1933, RMSE_E_per_atom=1.9 meV, RMSE_F=92.0 meV / A, RMSE_stress=5.8 meV / A^3
2024-07-10 13:04:53.261 INFO: Training complete
2024-07-10 13:04:53.262 INFO: Computing metrics for training, validation, and test sets
2024-07-10 13:04:53.263 INFO: Loading checkpoint: checkpoints/umbrella_run-123_epoch-1050.pt
2024-07-10 13:04:53.436 INFO: Loaded model from epoch 1050
2024-07-10 13:04:53.437 INFO: Evaluating train ...
2024-07-10 13:04:56.970 INFO: Evaluating Default ...
2024-07-10 13:04:57.095 INFO: Evaluating pbe_mp ...
2024-07-10 13:04:57.487 INFO: 
+-------------+---------------------+------------------+-------------------+---------------------------------------+
| config_type | RMSE E / meV / atom | RMSE F / meV / A | relative F RMSE % | RMSE Stress (Virials) / meV / A (A^3) |
+-------------+---------------------+------------------+-------------------+---------------------------------------+
|    train    |         6.1         |       36.6       |        2.62       |                  15.3                 |
|   Default   |         3.2         |       93.0       |        9.63       |                  6.0                  |
|    pbe_mp   |        115.3        |      188.8       |       174.42      |                  19.5                 |
+-------------+---------------------+------------------+-------------------+---------------------------------------+
2024-07-10 13:04:57.488 INFO: Saving model to checkpoints/umbrella_run-123.model
2024-07-10 13:04:57.711 INFO: Loading checkpoint: checkpoints/umbrella_run-123_epoch-1198_swa.pt
2024-07-10 13:04:58.205 INFO: Loaded model from epoch 1198
2024-07-10 13:04:58.206 INFO: Evaluating train ...
2024-07-10 13:05:01.775 INFO: Evaluating Default ...
2024-07-10 13:05:01.914 INFO: Evaluating pbe_mp ...
2024-07-10 13:05:02.325 INFO: 
+-------------+---------------------+------------------+-------------------+---------------------------------------+
| config_type | RMSE E / meV / atom | RMSE F / meV / A | relative F RMSE % | RMSE Stress (Virials) / meV / A (A^3) |
+-------------+---------------------+------------------+-------------------+---------------------------------------+
|    train    |         3.8         |       26.4       |        1.94       |                  15.3                 |
|   Default   |         1.9         |       92.0       |        9.52       |                  5.8                  |
|    pbe_mp   |        119.0        |      194.7       |       179.88      |                  20.3                 |
+-------------+---------------------+------------------+-------------------+---------------------------------------+
2024-07-10 13:05:02.326 INFO: Saving model to checkpoints/umbrella_run-123_swa.model
2024-07-10 13:05:02.522 INFO: Done
```
